File: web/packages/dtx-core/tests/record.test.ts
import { describe, it, expect } from 'vitest';
import { mergeChartRecord } from '../src/scoring/record.js';
import { Judgment, type JudgmentKind } from '../src/scoring/judgment.js';
import type { ScoreSnapshot } from '../src/scoring/score.js';

/** Build a minimal snapshot inline — the full tracker path is tested
 * elsewhere; we just need something mergeChartRecord can read. */
function snap(over: Partial<ScoreSnapshot>): ScoreSnapshot {
  const counts: Record<JudgmentKind, number> = {
    [Judgment.PERFECT]: 0,
    [Judgment.GREAT]: 0,
    [Judgment.GOOD]: 0,
    [Judgment.POOR]: 0,
    [Judgment.MISS]: 0,
  };
  return {
    totalNotes: 100,
    counts: { ...counts, ...(over.counts ?? {}) },
    combo: over.combo ?? 0,
    maxCombo: over.maxCombo ?? 0,
    score: over.score ?? 0,
    autoCount: over.autoCount ?? 0,
  };
}

describe('mergeChartRecord', () => {
  it('first play produces a record with plays=1', () => {
    const rec = mergeChartRecord(
      'Songs/a.dtx',
      null,
      snap({ score: 500_000, maxCombo: 80, counts: { [Judgment.PERFECT]: 80 } as Record<JudgmentKind, number> })
    );
    expect(rec.chartPath).toBe('Songs/a.dtx');
    expect(rec.plays).toBe(1);
    expect(rec.bestScore).toBe(500_000);
    expect(rec.lastPlayedMs).toBeGreaterThan(0);
  });

  it('score / achievement / rank take max across plays', () => {
    const first = mergeChartRecord('x.dtx', null, snap({ score: 400_000, counts: { [Judgment.PERFECT]: 60 } as Record<JudgmentKind, number>, maxCombo: 60 }));
    const second = mergeChartRecord('x.dtx', first, snap({ score: 700_000, counts: { [Judgment.PERFECT]: 90 } as Record<JudgmentKind, number>, maxCombo: 90 }));
    expect(second.bestScore).toBe(700_000);
    expect(second.plays).toBe(2);
    // Going backwards doesn't erase the best.
    const third = mergeChartRecord('x.dtx', second, snap({ score: 100_000, counts: { [Judgment.PERFECT]: 20 } as Record<JudgmentKind, number>, maxCombo: 20 }));
    expect(third.bestScore).toBe(700_000);
    expect(third.plays).toBe(3);
  });

  it('full-combo flag is sticky — once true, stays true', () => {
    const pristine = mergeChartRecord(
      'x.dtx',
      null,
      snap({ counts: { [Judgment.PERFECT]: 80, [Judgment.GREAT]: 20 } as Record<JudgmentKind, number>, maxCombo: 100 })
    );
    expect(pristine.fullCombo).toBe(true);
    // Play again and drop combo; flag should persist.
    const after = mergeChartRecord(
      'x.dtx',
      pristine,
      snap({ counts: { [Judgment.PERFECT]: 50, [Judgment.MISS]: 50 } as Record<JudgmentKind, number>, maxCombo: 40 })
    );
    expect(after.fullCombo).toBe(true);
  });

  it('excellent flag is sticky and implies full-combo on that play', () => {
    const perfect = mergeChartRecord(
      'x.dtx',
      null,
      snap({ counts: { [Judgment.PERFECT]: 100 } as Record<JudgmentKind, number>, maxCombo: 100 })
    );
    expect(perfect.excellent).toBe(true);
    expect(perfect.fullCombo).toBe(true);
    const worse = mergeChartRecord(
      'x.dtx',
      perfect,
      snap({ counts: { [Judgment.PERFECT]: 90, [Judgment.GREAT]: 10 } as Record<JudgmentKind, number>, maxCombo: 100 })
    );
    expect(worse.excellent).toBe(true);
  });

  it('rank only moves up, never down', () => {
    const B = mergeChartRecord(
      'x.dtx',
      null,
      snap({ counts: { [Judgment.PERFECT]: 70, [Judgment.GREAT]: 20, [Judgment.MISS]: 10 } as Record<JudgmentKind, number>, maxCombo: 70 })
    );
    const rankB = B.bestRank;
    const E = mergeChartRecord(
      'x.dtx',
      B,
      snap({ counts: { [Judgment.MISS]: 100 } as Record<JudgmentKind, number>, maxCombo: 0 })
    );
    expect(E.bestRank).toBe(rankB);
  });
});

File: native/quest-spatial/app/src/test/kotlin/com/dtxmania/quest/dtxcore/scoring/RecordTest.kt
package com.dtxmania.quest.dtxcore.scoring

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

/**
 * Ported 1:1 from `web/packages/dtx-core/tests/record.test.ts`.
 *
 * Note on time: the TS source calls `Date.now()` directly inside
 * [mergeChartRecord]; the Kotlin port adds an explicit `nowMs`
 * parameter (default `System.currentTimeMillis()`) so the test can
 * inject a deterministic value when it asserts on `lastPlayedMs`.
 */
class RecordTest {

    private fun snap(
        counts: Map<Judgment, Int> = emptyMap(),
        combo: Int = 0,
        maxCombo: Int = 0,
        score: Int = 0,
        autoCount: Int = 0,
    ): ScoreSnapshot {
        val full = mutableMapOf(
            Judgment.PERFECT to 0,
            Judgment.GREAT to 0,
            Judgment.GOOD to 0,
            Judgment.POOR to 0,
            Judgment.MISS to 0,
        )
        full.putAll(counts)
        return ScoreSnapshot(
            totalNotes = 100,
            counts = full,
            combo = combo,
            maxCombo = maxCombo,
            score = score,
            autoCount = autoCount,
        )
    }

    @Test fun `first play produces a record with plays equals 1`() {
        val rec = mergeChartRecord(
            "Songs/a.dtx",
            null,
            snap(score = 500_000, maxCombo = 80, counts = mapOf(Judgment.PERFECT to 80)),
            nowMs = 12345L,
        )
        assertEquals("Songs/a.dtx", rec.chartPath)
        assertEquals(1, rec.plays)
        assertEquals(500_000, rec.bestScore)
        assertTrue(rec.lastPlayedMs > 0)
    }

    @Test fun `score, achievement, rank take max across plays`() {
        val first = mergeChartRecord(
            "x.dtx", null,
            snap(score = 400_000, counts = mapOf(Judgment.PERFECT to 60), maxCombo = 60),
        )
        val second = mergeChartRecord(
            "x.dtx", first,
            snap(score = 700_000, counts = mapOf(Judgment.PERFECT to 90), maxCombo = 90),
        )
        assertEquals(700_000, second.bestScore)
        assertEquals(2, second.plays)
        // Going backwards doesn't erase the best.
        val third = mergeChartRecord(
            "x.dtx", second,
            snap(score = 100_000, counts = mapOf(Judgment.PERFECT to 20), maxCombo = 20),
        )
        assertEquals(700_000, third.bestScore)
        assertEquals(3, third.plays)
    }

    @Test fun `full-combo flag is sticky - once true, stays true`() {
        val pristine = mergeChartRecord(
            "x.dtx", null,
            snap(counts = mapOf(Judgment.PERFECT to 80, Judgment.GREAT to 20), maxCombo = 100),
        )
        assertTrue(pristine.fullCombo)
        // Play again and drop combo; flag should persist.
        val after = mergeChartRecord(
            "x.dtx", pristine,
            snap(counts = mapOf(Judgment.PERFECT to 50, Judgment.MISS to 50), maxCombo = 40),
        )
        assertTrue(after.fullCombo)
    }

    @Test fun `excellent flag is sticky and implies full-combo on that play`() {
        val perfect = mergeChartRecord(
            "x.dtx", null,
            snap(counts = mapOf(Judgment.PERFECT to 100), maxCombo = 100),
        )
        assertTrue(perfect.excellent)
        assertTrue(perfect.fullCombo)
        val worse = mergeChartRecord(
            "x.dtx", perfect,
            snap(counts = mapOf(Judgment.PERFECT to 90, Judgment.GREAT to 10), maxCombo = 100),
        )
        assertTrue(worse.excellent)
    }

    @Test fun `rank only moves up, never down`() {
        val b = mergeChartRecord(
            "x.dtx", null,
            snap(
                counts = mapOf(Judgment.PERFECT to 70, Judgment.GREAT to 20, Judgment.MISS to 10),
                maxCombo = 70,
            ),
        )
        val rankB = b.bestRank
        val e = mergeChartRecord(
            "x.dtx", b,
            snap(counts = mapOf(Judgment.MISS to 100), maxCombo = 0),
        )
        assertEquals(rankB, e.bestRank)
    }
}
